ethexe/consensus/src/utils.rs

```diff
@@ -0,0 +1,354 @@
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # Utilities Module
+//!
+//! This module provides utility functions and data structures for handling batch commitments,
+//! validation requests, and multi-signature operations in the Ethexe system.
+
+use anyhow::Result;
+use ethexe_common::gear::{BatchCommitment, BlockCommitment, CodeCommitment};
+use ethexe_signer::{
+    sha3::digest::Update, Address, ContractSignature, ContractSigner, Digest, PublicKey, ToDigest,
+};
+use gprimitives::H256;
+use parity_scale_codec::{Decode, Encode};
+use std::collections::BTreeMap;
+
+/// Represents a request for validating a batch of block commitments.
+/// This structure is used to verify the integrity and validity of multiple block commitments
+/// and their associated code commitments.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BatchCommitmentValidationRequest {
+    /// List of block commitment validation requests
+    pub blocks: Vec<BlockCommitmentValidationRequest>,
+    /// List of code commitments to be validated
+    pub codes: Vec<CodeCommitment>,
+}
+
+impl From<&BatchCommitment> for BatchCommitmentValidationRequest {
+    fn from(batch: &BatchCommitment) -> Self {
+        BatchCommitmentValidationRequest {
+            blocks: batch
+                .block_commitments
+                .iter()
+                .map(BlockCommitmentValidationRequest::from)
+                .collect(),
+            codes: batch.code_commitments.clone(),
+        }
+    }
+}
+
+impl ToDigest for BatchCommitmentValidationRequest {
+    fn update_hasher(&self, hasher: &mut ethexe_signer::sha3::Keccak256) {
+        hasher.update(self.codes.to_digest().as_ref());
+        hasher.update(self.blocks.to_digest().as_ref());
+    }
+}
+
+/// A request for validating a single block commitment.
+/// Contains all necessary information to verify the integrity of a block's commitment.
+///
+/// NOTE: [`BlockCommitmentValidationRequest`] digest is always equal to the corresponding
+/// [`BlockCommitment`] digest.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct BlockCommitmentValidationRequest {
+    /// Hash of the block being validated
+    pub block_hash: H256,
+    /// Timestamp of the block
+    pub block_timestamp: u64,
```

Comment:
```suggestion
    pub hash: H256,
    /// Timestamp of the block
    pub timestamp: u64,
```

Reply:
You can ask a trillion of meaningless questions about whatever exist on Earth. `block.block_hash` and `block.block_timestamp` seem excessive. For clarity you always have doc comment to address questions like hash of what etc

Reply:
not important